### PR TITLE
Collision API

### PR DESCRIPTION
**Pull request by Owen1212055**
This gives you the ability to check for collisions for an entity or a bounding box in a world. 
This allows you to see for example if an entity can fit at a certain location.

Feedback appreciated.

### CODE REVIEW

File: patches/api/0337-Collision-API.patch [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
++     *
++     * Checks for any collisions in the given boundingbox.
++     * This will check for any colliding entities / worldborder / blocks.
++     *
++     * @param boundingBox the box to check collisions in
++     * @param loadChunks should load chunks within bounding box
++     * @return collides or not
++     */
++    boolean collidesAt(@NotNull BoundingBox boundingBox, boolean loadChunks);
++    // Paper End - Collision API
```
Comment (NoahvdAa) [inline patches/api/0337-Collision-API.patch]: Put this in the existing paper block instead (line 37 of this patch)

File: patches/server/0819-Collision-API.patch [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+     }
+     // Paper end
++    // Paper Start - Collision API
++    @Override
++    public boolean collidesAt(@org.jetbrains.annotations.NotNull BoundingBox boundingBox, boolean loadChunks) {
++        net.minecraft.world.phys.AABB aabb = new AABB(boundingBox.getMinX(), boundingBox.getMinY(), boundingBox.getMinZ(), boundingBox.getMaxX(), boundingBox.getMaxY(), boundingBox.getMaxZ(), false);
++
++        return !getHandle().level.noCollision(getHandle(), aabb);
++    }
++    // Paper End - Collision API
```
Comment (NoahvdAa) [inline patches/server/0819-Collision-API.patch]: Same here, put this in the existing paper block.

File: patches/server/0819-Collision-API.patch [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
++++ b/src/main/java/org/bukkit/craftbukkit/CraftRegionAccessor.java
+@@ -894,4 +894,13 @@ public abstract class CraftRegionAccessor implements RegionAccessor {
+ 
+         throw new IllegalArgumentException("Cannot spawn an entity for " + clazz.getName());
+     }
++
++    // Paper Start - Collision API
++    @Override
++    public boolean collidesWith(@org.jetbrains.annotations.NotNull org.bukkit.util.BoundingBox boundingBox) {
++        net.minecraft.world.phys.AABB aabb = new AABB(boundingBox.getMinX(), boundingBox.getMinY(), boundingBox.getMinZ(), boundingBox.getMaxX(), boundingBox.getMaxY(), boundingBox.getMaxZ(), false);
```
Comment (NoahvdAa) [inline patches/server/0819-Collision-API.patch]: Maybe null check the bounding box? Something like
```java
Preconditions.checkArgument(boundingBox != null, "boundingBox cannot be null");
```

File: patches/server/0819-Collision-API.patch [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+@@ -1264,4 +1264,12 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
+         return getHandle().isTicking();
+     }
+     // Paper end
++    // Paper Start - Collision API
++    @Override
++    public boolean collidesAt(@org.jetbrains.annotations.NotNull BoundingBox boundingBox, boolean loadChunks) {
++        net.minecraft.world.phys.AABB aabb = new AABB(boundingBox.getMinX(), boundingBox.getMinY(), boundingBox.getMinZ(), boundingBox.getMaxX(), boundingBox.getMaxY(), boundingBox.getMaxZ(), false);
```
Comment (NoahvdAa) [inline patches/server/0819-Collision-API.patch]: Same null check here.

**Comment by Owen1212055** [comment]
Removed the chunk loading param and moved some more of the implementation into the craft classes. This should resolve the error with the check collisions at the location method. (The location was offset)

File: patches/server/0844-Collision-API.patch [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
++    @Override
++    public boolean collidesAt(@org.jetbrains.annotations.NotNull Location location) {
++        net.minecraft.world.phys.AABB aabb = this.getHandle().getBoundingBoxAt(location.getX(), location.getY(), location.getZ());
++
++        return !this.getHandle().level.noCollision(this.getHandle(), aabb);
++    }
++
++    @Override
++    public boolean collidesAt(@org.jetbrains.annotations.NotNull BoundingBox boundingBox) {
++        net.minecraft.world.phys.AABB aabb = new AABB(boundingBox.getMinX(), boundingBox.getMinY(), boundingBox.getMinZ(), boundingBox.getMaxX(), boundingBox.getMaxY(), boundingBox.getMaxZ(), false);
```
Comment (lynxplay) [inline patches/server/0844-Collision-API.patch]: In regards to repeated conversion between AABB and BoundingBox, the MCUtil utility class does not yet seem to have a neat conversion method for this.

Might be worth to add one [into this patch](https://github.com/PaperMC/Paper/blob/master/patches/server/0006-MC-Utils.patch) and use that to convert from/to an AABB.

Incredibly minor nitpick tho, so really up to you.

File: patches/api/0347-Collision-API.patch [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+      */
+     @NotNull
+     public <T extends Entity> T spawn(@NotNull Location location, @NotNull Class<T> clazz, boolean randomizeData, @Nullable Consumer<T> function) throws IllegalArgumentException;
++
++    // Paper Start - Collision API
++
++    /**
++     *
++     * Checks for any collisions in the given boundingbox.
++     * This will check for any colliding hard entities / worldborder / blocks.
```
Comment (lynxplay) [inline patches/api/0347-Collision-API.patch]: What exactly are "hard entities" ? Might need clarification that this mainly wraps boats, shulkers and minecarts (as of right now).

File: patches/server/0844-Collision-API.patch (hunk 5 above)
Comment (Owen1212055) [reply patches/server/0844-Collision-API.patch]: Decided against this in the end, as really this constructor is only needed here once anyways... So IDK, and the constructor used in question is added in a later patch.

File: patches/api/0347-Collision-API.patch (hunk 6 above)
Comment (Owen1212055) [reply patches/api/0347-Collision-API.patch]: Gave some examples!

**Comment by Owen1212055** [comment]
Rebased

**Comment by Owen1212055** [comment]
Rebased

**Review by kennytv — APPROVED** [review]

**Comment by Owen1212055** [comment]
Rebased and fixed documentation, I appreciate any last feedback but functionality seems to be fine.

**Comment by Owen1212055** [comment]
Rebased

**Review by kennytv — APPROVED** [review]

**Comment by Owen1212055** [comment]
Rebased

**Comment by Owen1212055** [comment]
Rebased

**Review by electronicboy — APPROVED** [review]

**Comment by Owen1212055** [comment]
image
A useful diagram that may or may not help people.

**Pull request merged**